File: Structural/Proxy/DIY.java
package Structural.Proxy;

import java.util.HashMap;
import java.util.Map;

public class DIY {
    public static void main(String[] args) {
        var playlist = new Playlist();
        String [] songs = {"The one that got away", "My Tears Rocochet", "Belly Ache"};
        for (var song : songs){
            playlist.add(new SongProxy(song));
        }
        playlist.openSong("The one that got away");
    }
}

//Real Object
class RealSong implements Song{
private String song;

    public RealSong(String song) {
        this.song = song;
        load();
    }
   private void load(){
        System.out.println("Buffering...");
   }

    @Override
    public void show() {
        System.out.println("Playing song: " + song);
    }

    @Override
    public String getSong() {
        return song;
    }
}

//Proxy - substitute for the real object
class SongProxy implements Song{
    private RealSong realSong;
    private String song;

    public SongProxy(String song) {
        this.song = song;
    }

    @Override
    public void show() {
        if (realSong == null){
        realSong = new RealSong(song);
        }

        realSong.show();
    }

    @Override
    public String getSong() {
      return song;
    }

}
//Library 

class Playlist {
    private Map<String,Song> songs = new HashMap<>();
    public void add(Song song){
        songs.put(song.getSong(), song);
    }
    public void openSong(String song){
        songs.get(song).show();
    }

    
}
interface Song{
    void show();
    String getSong();
}

File: TS/Structural/Proxy/DIY.ts
interface ProxySong {
  show(): void;
  getSong(): string;
}

// Real Object
class RealSong implements ProxySong {
  private song: string;

  constructor(song: string) {
    this.song = song;
    this.load();
  }

  private load(): void {
    console.log("Buffering...");
  }

  show(): void {
    console.log("Playing song: " + this.song);
  }

  getSong(): string {
    return this.song;
  }
}

// Proxy - substitute for the real object
class SongProxy implements ProxySong {
  private realSong: RealSong | null = null;
  private song: string;

  constructor(song: string) {
    this.song = song;
  }

  show(): void {
    if (this.realSong === null) {
      this.realSong = new RealSong(this.song);
    }
    this.realSong.show();
  }

  getSong(): string {
    return this.song;
  }
}

// Library
class Playlist {
  private songs: ProxySong[] = [];

  add(song: ProxySong): void {
    this.songs.push(song);
  }

  openSong(song: string): void {
    const foundSong = this.songs.find((s) => s.getSong() === song);
    foundSong?.show();
  }
}

class DIYProxy {
  public static main(): void {
    const playlist = new Playlist();
    const songs = ["The one that got away", "My Tears Ricochet", "Belly Ache"];
    for (const song of songs) {
      playlist.add(new SongProxy(song));
    }
    playlist.openSong("The one that got away");
  }
}

DIYProxy.main();
